// @mangar2/mqttclient/test/connect-cases.js
module.exports = [
    {
        description: 'Connect class - Successful connection',
        clientId: 'test-client',
        brokerHost: 'localhost',
        tests: [
            {
                description: 'Testing connect method with clean session',
                method: 'connect',
                args: [true, '1.0', 600000],
                resultKey: 'connectResult',
                expected: {
                    connectResult: {
                        // Add expected result structure here, for example:
                        token: { send: 'exampleTokenSend', receive: 'exampleTokenReceive' }
                    }
                }
            },
            {
                description: 'Testing subscribe method',
                method: 'subscribe',
                args: [{ 'test/topic': 1 }, '1.0'],
                resultKey: 'subscribeResult',
                expected: {
                    subscribeResult: {
                        // Add expected result structure for subscribe here
                    }
                }
            }
        ]
    },
    {
        description: 'Connect class - Disconnection and error handling',
        clientId: 'test-client',
        brokerHost: 'localhost',
        tests: [
            {
                description: 'Testing disconnect method',
                method: 'disconnect',
                args: ['1.0'],
                resultKey: 'disconnectResult',
                expected: {
                    disconnectResult: {
                        // Add expected result structure for disconnect here
                    }
                }
            },
            {
                description: 'Testing connect method with invalid broker',
                method: 'connect',
                args: [true, '1.0', 600000],
                resultKey: 'connectError',
                expected: {
                    connectError: {
                        error: 'Unable to connect status code: 404' // Example error message
                    }
                }
            }
        ]
    },
    // Add more test cases as needed
]
