// @mangar2/mqttclient/test/on-connect-cases.js
module.exports = [
    {
        description: 'Connect test',
        clientId: 'test-client',
        result: { clean: false, token: { send:'abcd', receive: 'efgh' }, present: 0, mqttcode: 0 },
        tests: [
            {
                description: 'Testing connect method, clean session',
                method: 'connect',
                args: ['1.0', { clean: true, port: 0, host: 'localhost' }],
                result: { 
                    statusCode: 200, 
                    headers: { 'content-type': 'application/json', packet: 'connack' }, 
                    payload: '{ "present": 0, "mqttcode": 0, "token": {"send": "abcd", "receive": "efgh" }}' 
                },
                expected: {
                    'result': {
                        clean: false,
                        present: 0,
                        mqttcode: 0,
                        token: { send: 'abcd', receive: 'efgh' }
                    },
                    'history': [
                        {
                            clean: true,
                            port: 0,
                            host: 'localhost',
                            clientId: 'test-client',
                            keepAlive: 0,
                            qos: 1,
                            version: '1.0'
                        }
                    ]
                }
            }
        ]
    }
]
