// @mangar2/types/types.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */

'use strict'

/**
 * @description
 * provides safe type checking functions: isFunction, isArray, isObject, isNull, isUndefined, isBoolean, isNumber, isString
 * @example
 * isFunction(text => console.log(text)) // returns true
 * isBoolean('true') // returns false
 */
const Types = {
    getType: (anything) => { return {}.toString.call(anything).slice(8, -1) },
    isFunction: (anything) => { return Types.getType(anything) === 'Function' },
    isArray: (anything) => { return Array.isArray(anything) },
    isObject: (anything) => { return Types.getType(anything) === 'Object' },
    isNull: (anything) => { return Types.getType(anything) === 'Null' },
    isUndefined: (anything) => { return Types.getType(anything) === 'Undefined' },
    isBoolean: (anything) => { return Types.getType(anything) === 'Boolean' },
    isNumber: (anything) => { return Types.getType(anything) === 'Number' },
    isString: (anything) => { return Types.getType(anything) === 'String' },
    isDate: (anything) => { return Types.getType(anything) === 'Date' }
}

module.exports = Types
