// @mangar2/mqtt-client/src/service/http-send-services.ts
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 * @overview Provides functions to connect to the (http based) yaha mqtt broker
 */

import { HttpClient } from '@mangar2/httpservice';
import { Publish } from '@mangar2/mqtt-client/src/service/publish';
import { Connect } from '@mangar2/mqtt-client/src/service/connect';

/**
 * Options for the broker
 */
export type BrokerOptions = {
    port: number;
    host: string;
}

export class HttpSendServices {
    private _httpClient: HttpClient; 
    private _publish: Publish;
    private _connect: Connect;

    /**
     * Creates a new http service
     * @param port the port on which the server listens
     */
    constructor(senderName:string, brokerOptions: BrokerOptions) {
        this._httpClient = new HttpClient(brokerOptions.host, brokerOptions.port, 'http');
        this._publish = new Publish();
        this._publish.on('send', (path, httpMethod, headers, payload) => {
            this._httpClient.sendv2({ path, method: httpMethod, headers, payload });
        })
        this._connect = new Connect(senderName);
        this._connect.on('send', (path, httpMethod, headers, payload ) => {
            this._httpClient.sendv2({ path, method: httpMethod, headers, payload });
        })
    }

    /**
     * Gets the publish object
     */
    get publish(): Publish {
        return this._publish;
    }

    /**
     * Gets the connect object
     */
    get connect(): Connect {
        return this._connect;   
    }

}
