// @mangar2/httpservice/httpclient.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 * @brief http/https client to send data to a server
 */

'use strict'

const http = require('http')
const https = require('https')
const Client = require('./client')

/**
 * @description
 * Class simplifying access to the node http service for http clients
 * @param {string} host host name or ip address
 * @param {string|number} port port number
 * @param {string} [type='http'] either 'http' or 'https' for http/https clients
 * @example
 * const client = new HTTPClient('myhost', 10000);
 * client.sendv2({ path: 'info/getdata/1', method: 'GET' })
 * client.post({ path: 'postdata/1', payload={info: 'hello world')}, type='json' })
 * client.get('info/getdata/1')
 */
class HTTPClient {
    constructor (host, port, type='http') {
        this._service = http
        if (type === 'https') {
            this._service = https
        } 
        this._client = new Client(host, port)
    }

    /**
     * @type Agent
     * Agent for http/https connections
     */
    set agent(agent) { this._client.agent = agent }
    get agent() { return this._client.agent }

    /**
     * Sets host name and port number
     * @param {string} host host name
     * @param {string} port port number
     */
    setConnection (host, port) {
        this._client.setConnection(host, port)
    }

    /**
     * Sends data. A payload of type "object" is automatically stringified, a string is not
     * @param {string} path http path
     * @param {string} method http send method PUT, GET, ...
     * @param {string|object} payload payload to send
     * @param {object} [headers={}] header to send
     * @returns {Promise<{statusCode, headers, payload}>} Promise; resolve = {statusCode, headers, payload}
     */
    send (path, method, payload, headers = {}) {
        return this._client.send(path, method, payload, headers, this._service)
    }
    /**
     * Sends data. Adds a content-type element to the header and stringifies the boy based on the type
     * @param {object} sendOptions all information required for sending
     * @param {string} sendOptions.path http path
     * @param {string} sendOptions.method http send method PUT, GET, ...
     * @param {object} [sendOptions.payload] payload to send
     * @param {object} [sendOptions.headers={}] header to send
     * @param {string} [type] type of the payload data: html, text, json, form, xml
     * @returns {Promise<{statusCode, headers, payload}>} Promise; resolve = {statusCode, headers, payload}
     */
    sendv2 (sendOptions) { return this._client.sendv2(sendOptions, this._service)}

    /**
     * Sends a post request
     * @param {object} sendOptions all information required for sending
     * @param {string} sendOptions.path http path
     * @param {object} [sendOptions.payload] payload to send
     * @param {object} [sendOptions.headers={}] header to send
     * @param {string} [type] type of the payload data: html, text, json, form, xml
     * @returns {Promise<{statusCode, headers, payload}>} Promise; resolve = {statusCode, headers, payload}
     */
    post(sendOptions) {
        sendOptions.method = 'POST'
        return this.sendv2(sendOptions)
    }

    /**
     * Sends a put request
     * @param {object} sendOptions all information required for sending
     * @param {string} sendOptions.path http path
     * @param {object} [sendOptions.payload] payload to send
     * @param {object} [sendOptions.headers={}] header to send
     * @param {string} [type] type of the payload data: html, text, json, form, xml
     * @returns {Promise<{statusCode, headers, payload}>} Promise; resolve = {statusCode, headers, payload}
     */    
    put(sendOptions) {
        sendOptions.method = 'PUT'
        return this.sendv2(sendOptions)
    }

    /**
     * Sends a get request
     * @param {string} path http path
     * @param {object} [headers={}] header to send
     * @returns {Promise<{statusCode, headers, payload}>} Promise; resolve = {statusCode, headers, payload}
     */
    getRequest(path, headers = {}) {
        return this.send(path, 'GET', '', headers)
    }

    /**
     * Aborts all open requests
     * @returns {promise}, resolved once all connections are closed
     */
    close () {
        return this._client.close()
    }
}

module.exports = HTTPClient
