// @mangar2/httpservice/dist/httpserver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTPServer = void 0;
const server_1 = __importDefault(require("./server"));
/**
 * Represents an HTTP server for handling HTTP requests.
 */
class HTTPServer {
    /**
     * Initializes a new instance of the HTTPServer.
     * @param {number} port The port on which the server will listen.
     */
    constructor(port) {
        this._server = new server_1.default(port);
    }
    /**
     * Gets the address information of the server.
     * @returns { AddressInfo | undefined } Address information or undefined if server is not started.
     */
    get address() {
        var _a;
        const addressInfo = (_a = this._server.server) === null || _a === void 0 ? void 0 : _a.address();
        if (typeof addressInfo === 'string' || addressInfo === null || addressInfo === undefined) {
            return undefined;
        }
        return addressInfo;
    }
    /**
     * Registers an event callback for the server.
     * @param {string} event The name of the event (e.g., 'GET', 'POST').
     * @param {HTTPCallback} callback The callback function to handle the event.
     */
    on(event, callback) {
        this._server.on(event, callback);
    }
    /**
     * Starts the HTTP server to listen for requests.
     */
    listen() {
        this._server.listen(false);
    }
    /**
     * Stops the server from accepting new connections and closes existing ones.
     * @returns {Promise<void>} A promise that resolves when the server is closed.
     */
    close() {
        return this._server.close();
    }
}
exports.HTTPServer = HTTPServer;
