// @mangar2/mqtt-client/dist/http-services/http-client-services.js
"use strict";
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 * @overview Provides functionalities to communicate with an mqtt broker over http
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = exports.getIPv4Address = void 0;
const os_1 = __importDefault(require("os"));
const utils_1 = require("@mangar2/utils");
const mqtt_utils_1 = require("@mangar2/mqtt-utils");
const index_1 = require("../index");
/**
 * @private
 * @description
 * Gets the IP v4 address of the current device
 * @returns {string} IP v4 address
 */
function getIPv4Address() {
    const networkInterfaces = os_1.default.networkInterfaces();
    for (const network of Object.values(networkInterfaces)) {
        if (!network)
            continue;
        for (const info of network) {
            if (info.family === 'IPv4' && !info.internal) {
                return info.address;
            }
        }
    }
    return 'localhost';
}
exports.getIPv4Address = getIPv4Address;
class HttpClient {
    constructor(options) {
        this.options = options;
        this.nextPacketId = 0;
        const { clientId, brokerOptions, listener, version } = options;
        this._server = new index_1.HttpReceiveServices(listener || 0);
        this._client = new index_1.HttpSendServices(brokerOptions);
        this._version = version || '1.0';
        this._clientId = clientId;
    }
    /**
     * @private
     * @description
     * Provides a packet id for the client
     * @returns {number} packet id
     */
    providePacketId() {
        return this.nextPacketId = (this.nextPacketId % 0xFFFF) + 1;
    }
    /**
     * Checks if the MQTT client supports a specific feature.
     * @param feature - The feature to check.
     * @returns True if the client supports the feature, false otherwise.
     */
    supportsFeature(feature) {
        if (feature === 'pingreq') {
            return this.options.version !== '0.0';
        }
        else if (feature in ['connect', 'disconnect', 'subscribe', 'publish']) {
            return true;
        }
        else {
            return false;
        }
    }
    /**
     * @private
     * @description
     * Extracts the error message from an error
     * @param {string|Error|any} err
     * @returns {string} error message
     */
    getMessage(err) {
        return utils_1.Types.isString(err) ? err : utils_1.Types.isError(err) ? err.message : 'unknown error';
    }
    /**
     * Connects to the MQTT server.
     *
     * @param clean - A boolean indicating whether to clean the session or not.
     * @returns A Promise that resolves to a ConnectResult object.
     * @throws An error if unable to connect.
     */
    async connect(clean) {
        try {
            const sendData = mqtt_utils_1.Interfaces.connect(this._version, {
                clientId: this._clientId,
                clean,
                port: this._server.port,
                host: getIPv4Address()
            });
            const result = await this._client.send('/connect', sendData.headers, sendData.payload);
            sendData.resultCheck(result);
            const {} = result;
            // resultCheck checked, that content type is application/json. So parsing should be ok.
            const parsedResult = JSON.parse(result.payload);
            const { present, token, mqttcode } = parsedResult;
            mqtt_utils_1.Logger.logger.log('out', `$SYS/${this._clientId}/connected`, `connected, ${present ? 'Session available, ' : ''} send-token:${token.send}, receive-token:${token.receive} ${mqttcode ? `, mqttcode: ${mqttcode}` : ''}`, 0);
            return parsedResult;
        }
        catch (err) {
            throw new Error(`Unable to connect: ${this.getMessage(err)}`);
        }
    }
    /**
     * Disconnects the MQTT client from the server.
     * @returns A promise that resolves when the client is successfully disconnected.
     * @throws An error if the disconnection fails.
     */
    async disconnect() {
        try {
            const sendData = mqtt_utils_1.Interfaces.disconnect(this._version, this.options.clientId);
            this._client.send;
            const result = await this._client.send('/disconnect', sendData.headers, sendData.payload);
            sendData.resultCheck(result);
            mqtt_utils_1.Logger.logger.log('out', `$SYS/${this._clientId}/disconnected`, `disconnected`);
        }
        catch (err) {
            throw new Error(`Unable to disconnect: ${this.getMessage(err)}`);
        }
    }
    /**
     * Subscribes to MQTT topics.
     *
     * @param topics - The topics to subscribe to.
     * @returns A promise that resolves to a SubscribeResult object.
     * @throws An error if unable to subscribe.
     */
    async subscribe(topics) {
        try {
            const sendData = mqtt_utils_1.Interfaces.subscribe(this._version, topics, this._clientId, this.providePacketId());
            const result = await this._client.send('/subscribe', sendData.headers, sendData.payload);
            sendData.resultCheck(result);
            mqtt_utils_1.Logger.logger.log('out', `$SYS/${this._clientId}/subscribe`, `subscribed successfully`);
            return JSON.parse(result.payload);
        }
        catch (err) {
            throw new Error(`Unable to subscribe: ${this.getMessage(err)}`);
        }
    }
    /**
     * Unsubscribes from MQTT topics.
     *
     * @param topics - The topics to unsubscribe from.
     * @returns A promise that resolves to an UnsubscribeResult object.
     * @throws An error if unable to unsubscribe.
     */
    async unsubscribe(topics) {
        try {
            const sendData = mqtt_utils_1.Interfaces.unsubscribe(this._version, topics, this._clientId, this.providePacketId());
            const result = await this._client.send('/unsubscribe', sendData.headers, sendData.payload);
            sendData.resultCheck(result);
            mqtt_utils_1.Logger.logger.log('out', `$SYS/${this._clientId}/unsubscribe`, `unsubscribed successfully`);
            return JSON.parse(result.payload);
        }
        catch (err) {
            throw new Error(`Unable to unsubscribe: ${this.getMessage(err)}`);
        }
    }
    /**
     * Sends a ping request to signal the client is alive
     * @returns {boolean} true on success
     */
    async pingreq(token) {
        try {
            const payload = { token };
            const headers = { 'content-type': 'application/json; charset=UTF-8' };
            const result = await this._client.send('/pingreq', headers, payload);
            if (result.statusCode !== 204) {
                throw new Error(`Illegal return code, expected 204 got ${result.statusCode}`);
            }
            if (result.headers.packet !== 'pingresp') {
                throw new Error(`Wrong packet id, expected \'pingresp\', got ${result.headers.packet}`);
            }
            ;
            mqtt_utils_1.Logger.logger.log('out', `$SYS/${this._clientId}/pingreq`, `pinged successfully`);
        }
        catch (err) {
            throw new Error(`Error in ping request: ${this.getMessage(err)}`);
        }
    }
    async publish(token, message, serviceName) {
        return this._client.publish(token, message, serviceName, this.options.version);
    }
    onPublish(callback) {
        this._server.onPublish.on('publish', callback);
    }
    start() {
        this._server.listen();
    }
    close() {
        this._server.close();
    }
}
exports.HttpClient = HttpClient;
