// @mangar2/rules/test/rule-class/check.js
module.exports = [
    {
        description: 'undefined member',
        ruleTree: {
        },
        tests: [
            {
                description: 'dontExists',
                set: { name: 'one', topic: 'topic', dontExist: 'anything' },
                expected: { 
                    rules: [ { topic: 'topic' } ], 
                    tree: { rules: { one: { topic: 'topic' } } },
                    invalid: [
                        { 
                            topic: 'topic', 
                            name: 'one', 
                            dontExist: 'anything', 
                            isValid: false, 
                            errors: { dontExist: 'unknown property, misspelled?'} 
                        } 
                    ] 
                }
            }
        ]
    },
    {
        description: 'variables',
        ruleTree: {
        },
        tests: [
            {
                description: 'internal variable',
                set: { name: 'one', topic: 'topic', value: '/time' },
                check: true,
                expected: { 
                    invalid: [],
                    usedVariables: { }
                }
            },
            {
                description: 'undefined variable',
                set: { name: 'one', topic: 'topic', value: 'variable/not/defined' },
                check: true,
                expected: { 
                    invalid: [],
                    usedVariables: { 'variable/not/defined': null }
                }
            },
            {
                description: 'defined variable',
                set: { name: 'one', topic: 'topic', value: ['+', 'variable/defined', 10] },
                variables: {
                    'variable/defined': 10
                },
                check: true,
                expected: { 
                    invalid: [],
                    usedVariables: { 'variable/defined': 10 }
                }
            }
        ]
    }
]
