// @mangar2/mqttclient/dist/service/publish.js
"use strict";
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Publish = void 0;
const utils_1 = require("@mangar2/utils");
const httpservice_1 = require("@mangar2/httpservice");
const mqtt_utils_1 = require("@mangar2/mqtt-utils");
class Publish {
    /**
     * Creates an instance of PublishMessage.
     * @param {string} host The MQTT host.
     * @param {number} port The MQTT port.
     * @param {Configuration} configuration Configuration for the publisher.
     */
    constructor(host, port, configuration) {
        this.topicQueues = {};
        this.nextPacketId = 1;
        this.terminate = false;
        this.client = new httpservice_1.HttpClient(host, port);
        this.retry = utils_1.Types.isNumber(configuration.retry) ? configuration.retry : 60;
    }
    /**
     * Provides the next packet ID for MQTT messages.
     * @returns {number} The next packet ID.
     */
    providePacketId() {
        this.nextPacketId = this.nextPacketId % 0xFFFF || 1;
        return this.nextPacketId++;
    }
    /**
     * Handles the PUBREL part of QoS 2 message flow.
     * @param {number} packetid The packet ID of the message.
     * @param {string} token Authentication token.
     * @param {string} topic The MQTT topic.
     * @returns {Promise<boolean>} True if PUBREL was successful, false otherwise.
     */
    async pubrel(packetid, token, topic) {
        const sendData = mqtt_utils_1.Interfaces.pubrel('1.0', token, packetid);
        for (let retryCount = 0; retryCount < this.retry; retryCount++) {
            this.topicQueues[topic].state = 'pubrel';
            try {
                const result = await this.client.send('/pubrel', 'PUT', sendData.payload, sendData.headers);
                if (sendData.resultCheck(result)) {
                    return true;
                }
            }
            catch (err) {
                // success === false; No need to change anything, retry
            }
            await (0, utils_1.delay)(1000 * Math.min(retryCount * retryCount, 60));
        }
        return false;
    }
    /**
      * Sends a message with appropriate retries.
      * @param {any} sendData Data to be sent.
      * @returns {Promise<boolean>} True if the message was successfully sent, throws otherwise.
      */
    async sendMessage(sendData) {
        const { headers, payload } = sendData;
        let result;
        let sendError = null;
        for (let retryCount = 0; retryCount < this.retry; retryCount++) {
            try {
                result = await this.client.send('/publish', 'PUT', payload, headers);
                const success = sendData.resultCheck(result);
                if (success) {
                    if (headers.qos === 2) {
                        return await this.pubrel(headers.packetid, payload.token, payload.message.topic);
                    }
                    return true;
                }
            }
            catch (err) {
                sendError = err;
                // success === false; No need to change anything, we retry
            }
            headers.dup = 1;
            await (0, utils_1.delay)(1000 * Math.min(retryCount ** 2, 60));
        }
        throw Error('Error sending Message: ' + (sendError ? sendError.message : 'Unknown error'));
    }
    /**
     * Publishes a message to an MQTT topic. For QoS 1 and QoS 2, messages are queued per topic and sent sequentially.
     * QoS 0 messages are sent without confirmation. For QoS 1 and QoS 2, this function manages the message flow,
     * ensuring that messages are sent in order and retried as necessary.
     *
     * @param {string} token Authentication token used for the MQTT connection.
     * @param {Message} message The message object to be published. This includes the topic, payload, and optionally QoS.
     * @param {string} [version='1.0'] The MQTT protocol version to use for the message. Defaults to '1.0'.
     * @returns {Promise<string[]>} A promise that resolves with the result of the publish operation. For QoS 0, it may resolve
     *                          immediately after sending. For QoS 1 and QoS 2, it resolves after the appropriate
     *                          acknowledgment process is completed. It returns a list of success/non success messages
     */
    async publish(token, message, version = '1.0') {
        var _a;
        var _b, _c;
        let result = [];
        const { qos = 1 } = message;
        if (qos === 0) {
            const sendData = mqtt_utils_1.Interfaces.publish(version, { token, message });
            const sendResult = await this.client.send('/publish', 'PUT', sendData.payload, sendData.headers).catch(() => {
                // Do not care, if send is successful for qos === 0. Still await is needed to safely catch rejects.
            });
            result.push(sendResult && sendResult.statusCode === 204 ? "send (qos 0)" : "not send (qos 0)");
        }
        else {
            // For qos > 0, messages are queued per topic and sent one at a time.
            (_a = (_b = this.topicQueues)[_c = message.topic]) !== null && _a !== void 0 ? _a : (_b[_c] = { state: 'ready', queue: [] });
            const topicQueue = this.topicQueues[message.topic];
            const sendData = mqtt_utils_1.Interfaces.publish(version, { token, message, dup: false, packetid: this.providePacketId() });
            topicQueue.queue.push(sendData);
            if (topicQueue.state === 'ready') {
                while (topicQueue.queue.length > 0) {
                    // state settings ensure that we only publish one message with qos > 0 at a time.
                    topicQueue.state = 'publish';
                    const firstElement = topicQueue.queue.shift();
                    const success = await this.sendMessage(firstElement);
                    result.push(success ? `delivered (qos ${firstElement.headers.qos})` : `not delivered(qos ${firstElement.headers.qos})`);
                    topicQueue.state = 'ready';
                }
            }
            else {
                result.push(`queued (qos ${qos})`);
            }
        }
        return result;
    }
    /**
     * Closes the MQTT client connection.
     * @returns {Promise<void>}
     */
    async close() {
        this.terminate = true;
        await this.client.close();
    }
}
exports.Publish = Publish;
