// @mangar2/mqtt-client/dist/test-service/registration.js
"use strict";
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2024 Volker Böhm
 * @overview
 * Provides registration for a test client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepare = void 0;
class TestService {
    constructor(config) {
        this.config = config;
        console.log(this.config);
    }
    handleMessage(message) {
        console.log(message);
        return [];
    }
    processTasks() {
        return { messages: [] };
    }
}
/**
 * Standard prepare function required for all services
 * @param {Object} config values configuration
 * @param {Values} [values=null] existing values class
 * @returns {Values} the prepared values object
 */
const prepare = (config, service = null) => {
    const result = service || new TestService(config);
    return result;
};
exports.prepare = prepare;
