// @mangar2/mqtt-client/dist/test-service/test-service-registration.js
"use strict";
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2024 Volker Böhm
 * @overview
 * Provides registration for a test client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = void 0;
const mqtt_utils_1 = require("@mangar2/mqtt-utils");
const register_service_1 = require("../mqtt-service/register-service");
const mqtt_client_1 = require("../mqtt-client/mqtt-client");
const http_client_services_1 = require("../http-services/http-client-services");
const httpClientOptions = {
    clientId: 'test-service',
    version: '1.0',
    brokerOptions: {
        port: 8183,
        host: '192.168.0.4'
    },
    listener: 0
};
const mqttClientOptions = {
    clientId: 'test-service',
    clean: true,
    keepAliveInSeconds: 60
};
const config = {
    runservices: {
        clientId: 'test-service',
        version: '1.0',
        clean: true,
        broker: {
            port: 8183,
            host: '192.168.0.4'
        },
        listener: 0,
        keepAliveInSeconds: 60,
        services: ['test-service'],
        log: []
    },
    'test-service': {
        intervalInSeconds: 10,
        topics: {
            'test/test-service/null': 0,
            'test/test-service/eins': 1,
            'test/test-service/zwei': 2
        }
    }
};
const run = async () => {
    const logger = new mqtt_utils_1.Logger();
    logger.changePattern([{ module: 'test-service', topic: '$SYS/#', level: 1 }]);
    const httpClient = new http_client_services_1.HttpClient(httpClientOptions, logger);
    const mqttClient = new mqtt_client_1.MqttClient(mqttClientOptions, httpClient, logger);
    await mqttClient.run();
    await (0, register_service_1.registerService)(mqttClient, config, 'test-service', "@mangar2/mqtt-client/dist/test-service/test-service");
};
exports.run = run;
