// @mangar2/motion/checkrule.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */

'use strict'

const CheckInput = require('@mangar2/checkinput')

const checkRule = new CheckInput({
    title: 'rules',
    type: 'object',
    properties: {
        name: { type: 'string' },
        durationWithoutMovementInMinutes: { type: 'number' },
        requireAll: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        requireAny: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        allow: {
            type: 'array',
            minItems: 1,
            items: { type: 'string' }
        },
        deny: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        topic: {
            type: ['string', 'array'],
            minItems: 1,
            items: { type: 'string' }
        },
        check: {
            type: ['number', 'string', 'array'],
            minItems: 2,
            maxItems: 3,
            anyOf: [
                {
                    items: [
                        {
                            enum: [
                                'and', 'or',
                                '=', 'equal', 'equals',
                                '!=', 'not equal', 'not equals',
                                '<', 'below', 'less',
                                '>', 'above', 'greater',
                                '<=', 'below or equal', 'less or equal',
                                '>=', 'above or equal', 'greater or equal'
                            ]
                        }
                    ],
                    additionalItems: { $ref: '#/properties/check' }
                },
                {
                    items: [
                        {
                            enum: ['switch']
                        },
                        {
                            $ref: '#/properties/check'
                        },
                        {
                            type: 'object',
                            additionalProperties: { type: ['string', 'number', 'boolean'] }
                        }
                    ]
                },
                {
                    items: [
                        {
                            enum: ['call']
                        },
                        {
                            type: 'string'
                        }
                    ]
                }
            ]
        },
        value: {
            oneOf: [
                {
                    type: ['string', 'number']
                },
                {
                    $ref: '#/check'
                }
            ]
        },
        time: {
            oneOf: [
                {
                    type: 'string'
                },
                {
                    $ref: '#/check'
                }
            ]
        },
        duration: { type: ['number', 'string'] },
        QoS: {
            type: 'number',
            minimum: 0,
            maximum: 2
        }
    },
    required: ['topic', 'value'],
    additionalProperties: false

})

module.exports = checkRule
