// @mangar2/motion/motion.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */
'use strict'

const DEBUG = false

const Message = require('@mangar2/message')
const MotionHistory = require('./motionhistory')
const CheckInput = require('@mangar2/checkinput')
const TopicMatch = require('@mangar2/topicmatch')
const { Rules, ProcessRule } = require('@mangar2/rules')
const checkRule = require('./checkrule.js')
const types = require('@mangar2/types')

const errorLog = (err) => { require('@mangar2/errorlog')(err, DEBUG) }

const checkConfiguration = new CheckInput({
    type: 'object',
    properties: {
        motionTopics: {
            type: 'array',
            items: { type: 'string' },
            default: ['+/+/+/motion sensor/detection state']
        },
        presenceTopic: { type: 'string' },
        rules: { type: ['string', 'object'] },
        additionalProperties: false
    },
    required: ['motionTopics']
})

/**
 * @param {Object} options
 * @param {Array} [options.motionTopics] list of topics to subscribe to see all motions
 * @param {string} [options.presenceVariable='$SYS/presence'] name of the variable to store the presence flag
 * @param {Object} rulesTree tree with 'rules' elements
 */
class Motion {
    constructor (options, rulesTree) {
        if (!types.isArray(options.motionTopics)) {
            options.motionTopics = ['+/+/+/motion sensor/detection state']
        }
        this._presenceTopic = types.isString(options.presenceTopic) ? options.presenceTopic : '$SYS/presence'
        checkConfiguration.throwOnValidationError(options, 'Configuration error:')
        this._motionTopics = options.motionTopics
        this._history = new MotionHistory()
        this._rules = new Rules(rulesTree, checkRule)
        this._processRule = new ProcessRule()
        if (options.presenceVariable) {
            this._setVariable(options.presenceVariable, 'initial')
        } else {
            this._setVariable('$SYS/presence', 'initial')
        }
    }

    /**
     * Sets a single variable (for example from a message)
     * @param {string} name name of the variable
     * @param {string|number} value value of the variable
     */
    _setVariable (name, value) { this._processRule.setVariable(name, value) }

    /**
     * Clears the motion history
     */
    _clearMotionHistory () {
        this._history.clear()
    }

    /**
     * Checks receives messages and adds them to motion and variable lists
     * @param {Message} message received message
     */
    processMessage (message) {
        const result = []
        const topicMatch = new TopicMatch(this._motionTopics)
        if (topicMatch.getFirstMatch(message.topic) !== undefined) {
            this._history.addMotion(message)
        }
        const topic = message.topic
        topic.replace('/set', '')
        this._setVariable(topic, message.value)

        if (topic === this._motionTopics) {
            const resultMessage = new Message(topic, message.value, message.reason)
            resultMessage.addReason('set command received by motion')
            result.push(resultMessage)
        }
    }

    /**
     * Get all the variables to subscribe to
     * @param {number} [QoS=1] quality of service
     * @returns {string[]} array of subscriptions
     */
    getSubscriptions (QoS = 1) {
        const variables = this.checkRules()
        const subscriptions = {}
        subscriptions[this._motionTopics] = QoS
        subscriptions['$SYS/+/set'] = QoS
        const topicMatch = new TopicMatch(this._motionTopics)
        for (const variable in variables) {
            const isNew = topicMatch.getFirstMatch(variable) === undefined
            if (variables[variable] === undefined && isNew) {
                subscriptions[variable] = QoS
            }
        }
        return subscriptions
    }

    /**
     * Checks all rules for errors, prints errors to console
     * @returns {Object} map of variables (variable name/variable value)
     */
    checkRules () {
        let result = {}
        if (this._rules.rules.length === 0) {
            errorLog('no valid rules')
        }
        for (const errors of this._rules.invalidRules) {
            console.log('Error in rule "' + errors.name + '": ' + JSON.stringify(errors.messages, null, 2))
        }
        for (const rule of this._rules.rules) {
            console.log('Checking rule ' + rule.name)
            try {
                const usedVariables = this._processRule.determineNeededVariables(rule)
                result = { ...result, ...usedVariables }
            } catch (err) {
                this._rules.invalidateRule(rule.name)
                err.message = 'Error in rule "' + rule.name + '": ' + err.message
                errorLog(err)
            }
        }
        return result
    }

    /**
     * Check that the latest move timestamp is long enough ago
     * @param {number} durationWithoutMovementInMinutes required amount of minutes passed since last move
     * @param {numer} latestMoveTimestamp timestamp of the last move in milliseconds
     * @param {Date} curDate current date
     * @returns {boolean} true, if the last move is long enough ago
     */
    _isLatestMoveLongEnoughAgo (durationWithoutMovementInMinutes, latestMoveTimestamp, curDate) {
        let result = true
        if (types.isNumber(durationWithoutMovementInMinutes)) {
            const noMoveTimespan = curDate.getTime() - latestMoveTimestamp
            const ONE_MINUTE = 60 * 1000
            result = noMoveTimespan > durationWithoutMovementInMinutes * ONE_MINUTE
        }
        return result
    }

    /**
     * Sets the internal variable values from messages to send
     * @param {Message[]} messages array of messages to send
     */
    _setVariablesFromOwnMessages (messages) {
        for (const message of messages) {
            this._setVariable(message.topic, message.value)
        }
    }

    /**
     * Loops through all rules and processes them
     * @param {Date} [date] current date/time
     * @returns {Message[]} list of messages to process
     */
    processRules (date = new Date()) {
        const result = { messages: [], usedVariables: {} }
        this._processRule.date = date
        const recentMotion = this._history.getLatestMotionRoomList()
        for (const rule of this._rules.rules) {
            try {
                if (this._isLatestMoveLongEnoughAgo(rule.durationWithoutMovementInMinutes, recentMotion.timestamp, date)) {
                    const checkResult = this._processRule.check(rule, recentMotion.motions)
                    const messages = checkResult.messages
                    if (types.isArray(messages) && (messages.length > 0) && types.isObject(messages[0])) {
                        result.messages = [...result.messages, ...messages]
                    }
                }
            } catch (err) {
                this._rules.invalidateRule(rule.name)
                err.message = 'Error in rule "' + rule.name + '": ' + err.message
                errorLog(err)
            }
        }
        if (result.messages.length > 0) {
            this._clearMotionHistory()
            this._setVariablesFromOwnMessages(result.messages)
        }
        return result
    }
}

module.exports = Motion
