// @mangar2/motion/motion.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */
'use strict'

const DEBUG = false

const MotionHistory = require('./motionhistory')
const CheckInput = require('@mangar2/checkinput')
const Rules = require('@mangar2/rules')
const errorLog = (err) => { require('@mangar2/errorlog')(err, DEBUG) }

const checkRule = new CheckInput({
    type: 'object',
    properties: {
        name: { type: 'string' },
        status: { type: 'string' },
        durationWithoutMovementInMinutes: { type: 'number' },
        require: {
            type: 'array',
            items: { type: 'string' }
        },
        allow: {
            type: 'array',
            items: { type: 'string' }
        },
        deny: {
            type: 'array',
            items: { type: 'string' }
        },
        topic: {
            anyOf: [
                { type: 'string' },
                {
                    type: 'array',
                    items: { type: 'string' }
                }
            ]
        },
        value: {
            anyOf: [
                { type: 'string' },
                { type: 'number' }
            ]
        },
        time: { type: 'string' },
        duration: { type: 'number' }
    },
    required: ['current', 'durationWithoutMovementInMinutes', 'topic', 'value'],
    strict: true

})

class Motion {
    constructor (rulesTree) {
        this._history = new MotionHistory()
        this.rules = new Rules(rulesTree, checkRule)
    }

    /**
     * Gets the list of rules
     * @returns {Object[]} list of rules
     */
    get rules () { return this._rules }

    /**
     * Sets the list of rules
     * @param {Object[]} rules list of rules to process for automation
     */
    set rules (rules) { this._rules = rules }

    /**
     * Checks all rules for errors
     * @returns {Object} map of variables (variable name/variable value)
     */
    checkRules () {
        let result = {}
        if (this._rules.rules.length === 0) {
            errorLog('no rules specified')
        }
        for (const errors of this._rules.invalidRules) {
            console.log('Error in rule "' + errors.name + '": ' + JSON.stringify(errors.messages, null, 2))
        }
        for (const rule of this._rules.rules) {
            console.log('Checking rule ' + rule.name)
            try {
                const usedVariables = this._processRule.neededVariables(rule)
                result = { ...result, ...usedVariables }
            } catch (err) {
                this._rules.invalidateRule(rule.name)
                err.message = 'Error in rule "' + rule.name + '": ' + err.message
                errorLog(err)
            }
        }
        return result
    }
}

module.exports = new Motion()
