// @mangar2/rules/checkrule.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */
'use strict'

const CheckInput = require('@mangar2/checkinput')

const checkRule = new CheckInput({
    type: 'object',
    properties: {
        name: { type: 'string' },
        durationWithoutMovementInMinutes: { type: 'number' },
        allOf: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        anyOf: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        allow: {
            type: 'array',
            minItems: 1,
            items: { type: 'string' }
        },
        noneOf: {
            type: ['array', 'string'],
            minItems: 1,
            items: { type: 'string' }
        },
        topic: { $ref: '#/definitions/topic' },
        check: { $ref: '#/definitions/check' },
        value: { $ref: '#/definitions/check' },
        time: {
            allOf: [
                {
                    type: ['string', 'array']
                },
                {
                    $ref: '#/definitions/check'
                }
            ]
        },
        duration: { type: ['number', 'string'] },
        QoS: {
            type: 'number',
            minimum: 0,
            maximum: 2
        }
    },
    required: ['topic'],
    additionalProperties: false,
    definitions: {
        topic: {
            if: {
                required: ['value']
            },
            then: {
                topic: {
                    type: ['string', 'array'],
                    minItems: 1,
                    items: { type: 'string' }
                }
            },
            else: {
                topic: {
                    type: ['object'],
                    minProperties: 1,
                    properties: {
                        additionalItems: { $ref: '#/properties/value' }
                    }
                }
            }
        },
        check: {
            type: ['number', 'string', 'array'],
            anyOf: [
                {
                    items: [
                        {
                            enum: ['and', 'or']
                        }
                    ],
                    additionalItems: { $ref: '#/definitions/check' }
                },
                {
                    items: [
                        {
                            enum: [
                                '=', 'equal', 'equals',
                                '!=', 'not equal', 'not equals',
                                '<', 'below', 'less',
                                '>', 'above', 'greater',
                                '<=', 'below or equal', 'less or equal',
                                '>=', 'above or equal', 'greater or equal'
                            ]
                        },
                        { $ref: '#/definitions/check' },
                        { $ref: '#/definitions/check' }
                    ]
                },
                {
                    items: [
                        {
                            enum: ['switch']
                        },
                        {
                            $ref: '#/definitions/check'
                        },
                        {
                            type: 'object',
                            additionalProperties: { type: ['string', 'number', 'boolean'] }
                        }
                    ],
                    additionalItems: false
                },
                {
                    items: [
                        {
                            enum: ['call']
                        },
                        {
                            type: 'string'
                        }
                    ]
                }
            ]
        }
    }

})

module.exports = checkRule
