// @mangar2/mqtt/mqtt.js
/**
 * ---------------------------------------------------------------------------------------------------
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * File:      mqttWebServer.js
 *
 * Author:      Volker Böhm
 * Copyright:   Volker Böhm
 * ---------------------------------------------------------------------------------------------------
 */

'use strict'

const Message = require('@mangar2/message')
const LogFilter = require('@mangar2/logfilter')
const Connections = require('@mangar2/connections')
const CheckInput = require('@mangar2/checkinput')
const errorLog = require('@mangar2/errorlog')
const mqttVersion = require('@mangar2/mqttversion')

/**
 * Checks the payload structure of a connect command
 */
const checkConnectPayload = new CheckInput({
    type: 'object',
    properties: {
        clientId: { type: 'string' },
        host: { type: 'string' },
        port: { type: 'number' },
        clean: { type: 'boolean' },
        keepAlive: { type: 'number' },
        password: { type: 'string' },
        user: { type: 'string' },
        will: {
            type: 'object',
            properties: ''
        }
    },
    required: ['clientId', 'host', 'port', 'clean']
})

/**
 * Checks the payload structure of a disconnect command
 */
const checkDisconnectPayload = new CheckInput({
    type: 'object',
    properties: {
        clientId: { type: 'string' }
    },
    required: ['clientId']
})

/**
 * Checks the structure of a subscribe and unsubscribe
 */
const checkSubscribePayload = new CheckInput({
    type: 'object',
    properties: {
        clientId: { type: 'string' },
        topics: { type: 'object' },
        subscribe: {
            type: 'object',
            properties: {
                QoS: { type: 'number', minimum: 0, maximum: 2 },
                topics: {
                    anyOf: [
                        { type: 'string' },
                        {
                            type: 'array',
                            items: { type: 'string' }
                        }
                    ]
                }
            }
        }
    },
    required: ['clientId']
})

/**
 * Gets the quality of service from headers
 * @param {object} headers headers with qos field
 * @returns {number} qos
 */
function getQos (headers) {
    const qos = Number(headers.qos)
    return qos === 1 ? 1 : (qos === 2 ? 2 : 0)
}

function getDup (headers) {
    return !!(headers.dup === '1' || headers.dup === 'true')
}

function getRetain (headers) {
    return !!(headers.retain === '1' || headers.retain === 'true')
}

module.exports = class Mqtt {
    constructor (configuration) {
        this.connections = new Connections(configuration)
        this.logFilter = new LogFilter()
        if (configuration.log !== undefined) {
            this.logFilter.changePattern(configuration.log)
        }
        this.connections.restoreFromFile()

        this.callbacks = {}
        this.on('login', (username, password) => { return true })
        this.on('authenticate', (token) => { return true })
        this.on('send', (host, port, path, payload, headers) => {})
    };

    /**
     * Sets a callback
     * @param {string} event RESTful http verb (POST, GET, PUT, PATCH, DELETE)
     * @param {function} callback(payload, headers, path, res)
     * where payload is the http payload, headers the http headers, res the result structure and path is
     * is the http path as string
     */
    on (event, callback) {
        const eventLowerCase = event.toLowerCase()
        if (typeof (callback) === 'function') {
            this.callbacks[eventLowerCase] = callback
        }
    }

    /**
     * Callcs a callback stored to an event, if registered
     * @param {string} event event for callback
     * @param  {...any} param parameters to pass to the callback
     */
    invokeCallback (event, ...param) {
        let result
        if (this.callbacks[event] !== undefined) {
            result = this.callbacks[event](...param)
        }
        return result
    }

    /**
     * Reads the value from a parsed payload (if available)
     * @param {object} parsedpayload parsed Message payload
     * @returns {object} the message value
     */
    static parseValue (value) {
        if (value === undefined) {
            value = ''
        }

        try {
            value = Number(value)
        } catch (err) {
        }

        return value
    }

    /**
     * Sets the topics to log. Uses the mqtt topic match strings with wildchards # and +
     * @param {object} payload {topic: type, topic: type, ... } type = "received, send, all"
     * @returns {object} {Content-Type, body, statusCode}
     */
    setLog (payload) {
        this.logFilter.changePattern(payload)
        const result = {
            headers: { 'Content-Type': 'application/json' },
            payload: JSON.stringify(this.logFilter.moduleFilter),
            statusCode: 200
        }
        return result
    }

    /**
     * Internal use only: publishes a message
     * @param {object} message message to publish
     * @param {number} qos quality of service
     * @param {boolean} dup duplicate flag
     * @param {boolean} retain retain flag
     */
    publish (message, qos, dup, retain) {
        if (this.logFilter.getLogLevel('received', message.topic) !== undefined) {
            const reason = message.reason === undefined ? '' : message.reason[0].message
            const dupStr = dup ? ',dup' : ''
            const retainStr = retain ? ',retain' : ''
            const valueStr = message.value === '' || message.value === undefined ? '' : '=' + message.value
            console.log('%s (qos%s%s%s) received %s%s [%s]', new Date().toLocaleString(), qos, dupStr, retainStr, message.topic, valueStr, reason)
        }
        this.connections.publishMessage(message, qos, retain)
    }

    /**
     * Receives a pubrel packages
     * @param {object} payload publish payload
     * @param {object} headers headers parameter {packetid}
     * @returns {object} {headers, payload, statusCode}
     */
    onPubrel (payload, headers) {
        const result = mqttVersion.onPubrel(headers)
        this.connections.onPubrel(payload.token, result.packetid)
        return result
    }

    /**
     * Receives a message to pubish
     * @param {object} payload publish payload
     * @param {object} headers header parameters {qos, dup, packetid, retain}
     * @param {number} headers.qos quality of service
     * @param {(number|string)} [headers.dup] duplicate flag
     * @param {number} headers.packetid id of the packet
     * @param {(number|boolean)} [headers.retain] retain flag (true or 1 to retain the message)
     * @param {string} [headers.version='0.0'] interface version
     * @returns {object} {headers, payload, statusCode}
     */
    onPublish (payload, headers) {
        const token = payload.token
        const message = payload.message === undefined ? payload : payload.message
        const messageObj = new Message(message.topic, message.value, message.reason)
        let isDuplicate = false
        messageObj.addReason('received by broker')

        if (headers.version !== undefined && headers.version !== '0.0') {
            const clientId = this.connections.checkToken(token)
            isDuplicate = this.connections.isDuplicateQos2message(clientId, message.topic,
                getQos(headers), getDup(headers), headers.packetid)
        }

        const result = mqttVersion.onPublish(headers)

        if (!isDuplicate) {
            this.publish(messageObj, getQos(headers), getDup(headers), getRetain(headers), result.packetid)
        }

        return result
    }

    /**
     * Connects a client to the broker, storing the connection informations
     * @param {object} payload payload {clientId, host, port, clean, will}
     * @param {string} payload.clientId id of the connecting client
     * @param {string} payload.host host name (or ip) of the client
     * @param {number} payload.port port number
     * @param {boolean} payload.clean true, if the session shall be cleaned
     * @param {number} payload.keepAlive timeout in milliseconds for disconnecting
     * if no message is transmitted
     * @param {string} payload.password connection password
     * @param {string} payload.user connection user name
     * @param {object} payload.will message to be send on connection loss
     * @param {object} headers received headers
     * @param {string} headers.version interface version
     * @returns {{headers, payload, statusCode}}
     */
    onConnect (payload, headers) {
        checkConnectPayload.throwOnValidationError(payload, 'Connect wrong parameters: ')
        let result
        const { clientId, host, port, clean, keepAlive, password, user, will = {} } = payload
        const { version = '0.0' } = headers
        try {
            if (!this.invokeCallback('login', user, password)) {
                throw Error('Could not login, request from ' + payload.clientId)
            }
            const connect = this.connections.connect(clientId, host, port, clean, version, keepAlive)
            if (connect.mqttcode !== 0) {
                throw Error('Could not connect (code: ' + connect.mqttcode + ') request from ' + payload.clientId)
            }
            result = mqttVersion.onConnect(headers, { present: connect.present, token: connect.token })
            const message = new Message('$SYS/broker/connect/success', 1, 'request from ' + payload.clientId)
            this.connections.setWill(clientId, will)

            this.publish(message, 1, false, false)
        } catch (err) {
            const message = new Message('$SYS/broker/connect/refused', 1, err.message)
            this.publish(message, 1, false, false)
            throw err
        }
        return result
    }

    /**
     * disconnects a client from the broker
     * @param {object} payload
     * @param {string} payload.clientId Id of the client to disconnect
     * @param {object} headers received headers
     * @param {string} headers.version interface version
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    onDisconnect (payload, headers) {
        checkDisconnectPayload.throwOnValidationError(payload, 'Disconnect wrong parameters: ')
        this.connections.disconnect(payload.clientId)

        const message = new Message('$SYS/broker/disconnect/success', '', 'request from ' + payload.clientId)
        this.publish(message, 1, false, false)
        const result = mqttVersion.onDisconnect(headers)

        return result
    }

    /**
     * Subscribes to topics
     * @param {object} payload {clientId, topics}
     * @param {object} headers received headers
     * @param {number} headers.packetid id of the packet
     * @param {string} headers.version interface version
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    onSubscribe (payload, headers) {
        checkSubscribePayload.throwOnValidationError(payload, 'Subscribe wrong parameters: ')

        const qos = this.connections.subscribe(
            payload.clientId,
            payload.subscribe === undefined ? payload.topics : payload.subscribe
        )

        const message = new Message('$SYS/broker/subscribe', JSON.stringify(qos), 'request from ' + payload.clientId)
        this.publish(message, 1, false, false)

        const result = mqttVersion.onSubscribe(headers, qos)
        return result
    }

    /**
     * Unsubscribes to topics
     * @param {object} payload {clientId, topics}
     * @param {string} payload.clientId id of the connecting client
     * @param {object} payload.topics topics to subscribe to
     * @param {object} headers received headers
     * @param {string} headers.version interface version
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    onUnsubscribe (payload, headers) {
        checkSubscribePayload.throwOnValidationError(payload, 'Unsubscribe wrong parameters: ')

        const message = new Message('$SYS/broker/unsubscribe', '', 'request from ' + payload.clientId)
        this.publish(message, 1, false, false)

        this.connections.unsubscribe(payload.clientId, payload.topics)
        const result = mqttVersion.onUnsubscribe(headers)

        return result
    }

    /**
     * Answers to a ping
     * @param {object} payload http call payload
     * @param {string} playload.token
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    onPingreq (payload) {
        this.connections.getClientByToken(payload.token).actualizeLastActiveTimestamp()
        const message = new Message('$SYS/broker/ping', '', 'request from ' + payload.clientId)
        this.publish(message, 1, false, false)

        const result = {
            headers: { 'Content-Type': 'application/json', packet: 'pingresp' },
            payload: '',
            statusCode: 204
        }
        return result
    }

    /**
     * Calls a connection interface
     * @param {string} controls name of the mqtt controls to call (connect, disconnect, ...)
     * @param {object} payload parameters for the interface (see interfaces)
     * @param {object} headers headers of the http message
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    processRequest (controls, payload, headers) {
        let result = { headers: { 'content-type': 'text/plain' }, payload: 'request path not found: ' + controls, statusCode: 404 }
        try {
            switch (controls) {
            case 'publish': result = this.onPublish(payload, headers); break
            case 'pubrel': result = this.onPubrel(payload, headers); break
            case 'connect': result = this.onConnect(payload, headers); break
            case 'disconnect': result = this.onDisconnect(payload, headers); break
            case 'subscribe': result = this.onSubscribe(payload, headers); break
            case 'unsubscribe': result = this.onUnsubscribe(payload, headers); break
            case 'pingreq': result = this.onPingreq(payload); break
            case 'log': result = this.setLog(payload); break
            }
        } catch (err) {
            let error = (typeof (err) === 'string') ? err : err.message
            error = controls + ' ' + error
            errorLog(err)
            result = {
                headers: { 'content-type': 'application/json' },
                payload: JSON.stringify({ error }),
                statusCode: 400
            }
        }
        return result
    }

    /**
     * Gets all connections or a single connection (depends on path)
     * @param {string} clientId id of the client or ""
     * @returns {{headers: object, payload: object, statusCode: number}} http result objects
     */
    getConnections (clientId) {
        const result = {
            headers: { 'Content-Type': 'application/json' },
            payload: {},
            statusCode: 200
        }
        if (clientId === '') {
            result.payload = this.connections.clients
        } else {
            try {
                result.payload = this.connections.getClient(clientId, false)
            } catch (err) {
                result.payload = err.message
            }
        }
        return result
    }

    /**
     * Sends a publish message by invoking a callback
     * @param {MessageQueueEntry} options options for the callback
     * @param {string} options.host host name
     * @param {number} options.port port number
     * @param {string} options.version version string
     * @param {string} options.token send token
     * @param {Message} options.payload payload to send
     * @param {number} options.qos quality of service
     * @param {number} options.dup 1, if the packet is a duplicate (else 0)
     * @param {number} options.packetid packet id
     * @param {function} callback (host, port, payload, headers)
     * @returns {boolean} true, if sending was successful
     */
    async _sendPublish (options) {
        const { host, port, version, token, payload: message, qos, dup, packetid } = options
        const dupStr = dup ? ',dup' : ''
        const valueStr = message.value === '' || message.value === undefined ? '' : '=' + message.value
        const retain = false
        const publishData = mqttVersion.publish(version, token, message, qos, dup, retain, packetid)
        let success = false

        try {
            const result = await this.invokeCallback('send', host, port, 'publish', publishData.payload, publishData.headers)
            success = publishData.resultCheck(result)
            if (success) {
                this.connections.acknowledgeMessage(options)
            }
            if (this.logFilter.getLogLevel('send', message.topic) !== undefined) {
                const reason = message.reason === undefined ? '' : message.reason[0].message
                console.log('%s %s:%s (qos%s%s) send %s=%s [%s]', new Date().toLocaleString(), host, port, qos, dupStr, message.topic, valueStr, reason)
            }
        } catch (err) {
            if (this.logFilter.getLogLevel('send', '$SYS/broker/error') !== undefined) {
                console.log('%s %s:%s send %s=%s [%s]', new Date().toLocaleString(), host, port, message.topic, valueStr, err.message)
            }
        };
        return success
    }

    /**
     * Sends a pubrel message by invoking a callback
     * @param {MessageQueueEntry} options options for the callback
     * @param {string} options.host host name
     * @param {number} options.port port number
     * @param {string} options.version version string
     * @param {string} options.token send token
     * @param {number} options.packetid packet id
     * @param {function} callback (host, port, payload, headers)
     * @returns {boolean} true, if sending was successful
     */
    async _sendPubrel (options) {
        const { host, port, version, token, packetid } = options
        const publishData = mqttVersion.pubrel(version, token, packetid)
        let success = false

        try {
            const result = await this.invokeCallback('send', host, port, 'pubrel', publishData.payload, publishData.headers)
            success = publishData.resultCheck(result)
            if (success) {
                this.connections.acknowledgeMessage(options)
            }
        } catch (err) {
            if (this.logFilter.getLogLevel('send', '$SYS/broker/error') !== undefined) {
                console.log('%s %s:%s pubrel [%s]', new Date().toLocaleString(), host, port, err.message)
            }
        };
    }

    /**
     * Processes all messages to send and for each message calls the provided callback function.
     * @returns {number} amount of messages sent
     */
    async processSendMessage () {
        const messages = this.connections.getAllMessagesToSend()
        this.connections.cleanup()
        let promises = []
        for (const message of messages) {
            let result
            if (message.isStatusPubrel()) {
                result = this._sendPubrel(message)
            } else {
                result = this._sendPublish(message)
            }
            promises.push(result)
            if (promises.length >= 10) {
                Promise.all(promises)
                promises = []
            }
        }
        Promise.all(promises)
        return messages.length
    }
}
