// @mangar2/utils/src/index.ts
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 * @overview
 * This package combines a list of utlities needed in JavaScript development. The are created
 * in development of the yaha home automation system
 * - errorLog a function to write error messages to console (including timestamp)
 * - types a utilit to check the types of elements
 * - Callbacks a utility to manage callbacks in classes
 * - Retry a utility to call functions in a loop with a delay inbetween
 */


export { errorLog } from './errorlog';
export { Types } from './types';

export { Callbacks } from './callbacks';
export { Retry } from './retry';
export { delay } from './delay';
export { TaskQueue } from './taskqueue';
export { shutdown } from './shutdown';
export { sunset, sunrise } from './sun';
export { Persist } from './persist';
export { readConfiguration, selectConfiguration, getEnvironment, getCommandLineParameters } from './config'

export { deepMerge, StringIndexed } from './deep-merge';
