// @mangar2/checkinput/checkresult.js
/**
 * @private
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */

'use scrict'

const types = require('@mangar2/types')

/**
 * @private
 * @description Creates a result object
 * @param {boolean} check result of the check
 * @param {string} [message=''] initial message
 */
class CheckResult {
    constructor (check, message = '') {
        this.check = check
        this.messages = message
    }

    /**
     * @description Gets the check result
     * @returns {boolean}
     */
    get check () { return this._check }

    /**
     * Sets the check result
     * @param {boolean}  check
     */
    set check (check) { this._check = check }

    /**
     * @description Gets the messages
     * @returns {Object|string}
     */
    get messages () { return this._messages }

    /**
     * @description Sets the messages
     * @param {Object|string} messages
     */
    set messages (messages) { this._messages = messages }

    /**
     * @private
     * @description Adds a message to the message list
     * @param {string} message message to add
     */
    _addMessage (message) {
        if (this.messages === '') {
            this.messages = message
        } else if (types.isObject(this.messages)) {
            if (types.isString(message)) {
                this.messages.baseSchema = message
            } else {
                this.messages = { ...this.messages, ...message }
            }
        } else {
            if (!types.isArray(this.messages)) {
                this.messages = [this.messages]
            }
            if (types.isArray(message)) {
                this.messages.push(...message)
            } else {
                this.messages.push(message)
            }
        }
    }

    /**
     * @description Invalidates the result. It sets 'check' to false and the error message
     * @param {string|Object} message error message
     * @param {string} [property] name of the property causing the invalidation
     */
    invalidate (message, property) {
        if (message !== undefined && message !== '') {
            if (property === undefined) {
                this._addMessage(message)
            } else {
                if (!types.isObject(this.messages)) {
                    this.messages = {}
                }
                this.messages[property] = message
            }
        }
        this.check = false
    }

    /**
     * @description Sets the property related to the message
     * @param {string} property name of the property related to this message
     */
    setProperty (property) {
        const messages = this.messages
        this.messages = {}
        this.messages[property] = messages
    }

    /**
     * @description Adds an alternative result. the value is combined with an or. If value is still false
     * the messages are added in an array
     * @param {CheckResult} result result of an alternative evaluation
     */
    addAlternative (result) {
        this.check = this.check || result.check
        if (!this.check) {
            this.invalidate(result.messages)
        }
    }

    /**
     * @description Adds another check. the value is combined with an and, the messages are combined
     * @param {CheckResult} result result of an alternative evaluation
     * @param {propertyName} [propertyName] name of the property
     */
    addCheck (result, propertyName) {
        if (result.check !== undefined) {
            this.check = this.check && result.check
            if (!this.check) {
                this.invalidate(result.messages, propertyName)
            }
        }
    }
}

module.exports = CheckResult
