// @mangar2/mqtt-utils/src/mqtt-interface/connect.ts
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */

import { Types } from "@mangar2/utils";
import { IResult, RequestDataV2, standardHeaderJSON } from "./interfaces";

/**
 * Type representing the connection options
 */
export interface IConnectOptions {
    clientId: string;
    host: string;
    port: number;
    clean: boolean;
    keepAlive?: number; // keepalive time in seconds
}

/**
 * Type representing the onConnect options
 * @private
 * @interface
 * @param { number } present - 1 if the client is already present
 * @param { string } token - token for the client
 */
export interface IOnConnectOptions {
    present: number;
    token: string;
}


/**
 * Connection function interface.
 */
interface ConnectionFunction {
    (options: IConnectOptions): RequestDataV2
}

/**
 * Connect functions mapped by their version numbers
 * @private
 */
export const connect: Record<string, ConnectionFunction> = {
    '0.0': (options: IConnectOptions) => {
        const { clientId, host, port, clean } = options;
        const payload = {
            clientId, host, port, clean
        };
        const headers = { ...standardHeaderJSON, version: '0.0' };

        const resultCheck = (result: IResult) => {
            const success = result.statusCode === 200 &&
                Types.isString(result.headers['content-type']) &&
                result.headers['content-type'].startsWith('text/plain') &&
                result.payload.toLowerCase() === 'connack';
            return { success };
        }

        return { headers, payload, resultCheck };
    },

    '1.0': (options: IConnectOptions) => {
        const { clientId, host, port, clean, keepAlive = 0 } = options;
        const payload = {
            clientId, host, port, clean, keepAlive
        };
        const headers = { ...standardHeaderJSON, version: '1.0' };

         /**
         * Checks the result of the publish operation.
         * @param {IResult} result The result to be checked.
         * @returns {{success: boolean, message?:string}} success true on success, else false with error message
         */
         const resultCheck = (result: IResult) => {
            
            if (result.statusCode !== 200) {
                return { success: false, message: `status code 200 expected, got ${result.statusCode}`}
            };
            if (!Types.isString(result.headers['content-type']) || !result.headers['content-type']?.startsWith('application/json')) {
                return { success: false, message: 'returned content type is not application/json'}
            }
            const packet = result.headers.packet;
            if (result.headers.packet !== 'connack') {
                return { success: false, message: `acknowledge 'connack' expected, got ${packet}` }
            }
            if (!Types.isString(result.payload)) {
                return { success: false, message: 'payload not provided'}
            }

            return { success: true };
        }

        return { headers, payload, resultCheck };
    }
}

/**
 * Connection response interface.
 */
interface ConnectionResponseFunction {
    (payload: IOnConnectOptions): IResult
}

/**
 * OnConnect functions mapped by their version numbers
 * @private
 */
export const onConnect: Record<string, ConnectionResponseFunction> = {
    '0.0': () => {
        return {
            headers: { 'content-type': 'text/plain; charset=UTF-8', version: '0.0' },
            payload: 'connack',
            statusCode: 200
        };
    },

    '1.0': (payload: IOnConnectOptions) => {
        return {
            headers: { 'content-type': 'application/json; charset=UTF-8', packet: 'connack', version: '1.0' },
            payload: JSON.stringify(payload),
            statusCode: 200
        };
    }
}
