// @mangar2/mqtt-utils/src/mqtt-interface/interfaces.ts
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */


/**
 * Type representing the result object for interfaces
 */
export interface Result {
    statusCode: number;
    headers: Record<string, string>;
    payload: string;
    packetid?: number;
}

/**
 * Type representing the information returned for requests
 */
export interface RequestData {
    headers: Record<string, string>,
    payload: Record<string, any>,
    resultCheck: (result: Result) => boolean
}

/**
 * Standard header components for application/json messages
 */
export const standardHeaderJSON: Record<string, string> = {
    'content-type': 'application/json; charset=UTF-8',
    accept: 'application/json,text/plain',
    'accept-charset': 'UTF-8'
};

/**
 * Standard header components for text/plain messages
 */
export const standardHeaderText: Record<string, string> = {
    'content-type': 'text/plain; charset=UTF-8',
    accept: 'application/json,text/plain',
    'accept-charset': 'UTF-8'
};

/**
 * Supported types for ids
 */
export type idType = string | number;
