// httpclient/httpclient.js
/**
 * ---------------------------------------------------------------------------------------------------
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * File:      httpclient.js
 * Purpouse:
 * Sends a http request to a http server
 * Usage client = new HttpClient(host, port)
 * client.send(path, method, payload, headers)
 * 
 * Author:      Volker Böhm
 * Copyright:   Volker Böhm
 * Version:     1.0
 * ---------------------------------------------------------------------------------------------------
 */

'use strict';

const http = require('http');
const errorLog = require('@mangar2/errorlog');



module.exports = class HttpClient {

    constructor(host, port) {
        this.setConnection(host, port);
        this.callbacks = {};
        this.requests = {};
        this.nextRequestId = 0;
        
        this.on("error", (err) => {
            errorLog(err);
        });
    }

    /**
     * Sets host name and port number
     * @param {string} host host name
     * @param {string} port port number
     */
    setConnection(host, port) {

        if (host === undefined) {
            throw Error("No host specified");
        }
        if (port === undefined) {
            throw Error("No port specified");
        }

        this.host = host;
        this.port = port;
    }

    /**
     * Registeres callbacks. Supported:
     * result: Result
     * error: (error)
     * @param {string} event name of the event
     * @param {function} callback function callback
     */
    on(event, callback) {
        if (typeof(event) !== 'string') {
            throw Error("illegal event type in 'on'");
        }
        if (typeof(callback) !== 'function') {
            throw Error("callback is not a function");
        }
        this.callbacks[event] = callback;
    }

    /**
     * Calls a callback identified by the event name
     * @param {string} event name of the event (function to call)
     */
    call(event, ...param) {
        if (this.callbacks[event] !== undefined) {
            this.callbacks[event](...param);
        }
    }

    /**
     * Sends data
     * @param {string} path http path
     * @param {string} method http send method PUT, GET, ...
     * @param {object} payload payload to send
     * @param {object} headers header to send
     */
    send(path, method, payload, headers) {
        
        let stringPayload = JSON.stringify(payload);
        headers['Content-Length'] = stringPayload.length;

        let options = {
            host: this.host,
            port: this.port,
            path,
            method,
            headers
        };

        try {
            let requestId = this.nextRequestId;
            this.nextRequestId++;
            
            let request = http.request(options, res => {
                payload = "";
                res.setEncoding('utf8');
                res.on('data', chunk => payload += chunk);

                res.on('end', () => {
                    this.call('result', res.statusCode, res.headers, payload);
                    delete this.requests[requestId];
                });
            });
            this.requests[requestId] = request;
            request.write(stringPayload);
            request.end();
            request.on('error', err => {
                this.call("error", err);
                delete this.requests[requestId];
            });

        } catch (err) {
            this.call("error", err);
        }

    };

    /**
     * Aborts all open requests
     */
    close() {
        let result = new Promise((resolve, reject) => {
            for (let requestId in this.requests) {
                this.requests[requestId].abort();
            }
            resolve();
        })
        return result;
    }
}
