// @mangar2/valueservice/configuration.js
/**
 * @license
 * This software is licensed under the GNU LESSER GENERAL PUBLIC LICENSE Version 3. It is furnished
 * "as is", without any support, and with no warranty, express or implied, as to its usefulness for
 * any purpose.
 *
 * @author Volker Böhm
 * @copyright Copyright (c) 2020 Volker Böhm
 */

'use strict'

const sanitize = require('@mangar2/configuration')
const CheckInput = require('@mangar2/checkinput')


/**
 * JSON schema to check configuration input
 */
const checkConfiguration = new CheckInput({
    title: 'Value setting configuration',
    type: 'object',
    properties: {
        valuesFileName: {
            title: 'file holding the variables and their values',
            type: 'string'
        },
        subscribeQoS: { enum: [0, 1, 2] },
        intervalInSeconds: {
            description: 'Iinterval between checks for events based on rules. Additionally all rules will be checked on every receive Message',
            type: 'integer'

        },
        additionalProperties: false
    },
    required: ['valuesFileName', 'intervalInSeconds', 'subscribeQoS']
})

/**
 * Default values
 */
const defaultConfiguration = {
    valuesFileName: 'values.json',
    intervalInSeconds: 60,
    subscribeQoS: 1
}

/**
 * Checks the configuration and sets default values
 */
module.exports = (configuration) => {
    return sanitize(configuration, defaultConfiguration, checkConfiguration)
}
